fedbiomed/researcher/federated_workflows/_federated_workflow.py

```diff
@@ -502,37 +512,44 @@ def set_training_data(
             FedbiomedValueError : attempting to set training data from tags but self._tags is None
         """
         if not isinstance(from_tags, bool):
-            msg = ErrorNumbers.FB410.value + f' `from_tags` : got {type(from_tags)} but expected a boolean'
+            msg = ErrorNumbers.FB410.value + \
+                f' `from_tags` : got {type(from_tags)} but expected a boolean'
             logger.critical(msg)
             raise FedbiomedTypeError(msg)
         if from_tags and training_data is not None:
-            msg = ErrorNumbers.FB410.value + f' set_training_data: cannot specify a training_data argument if ' \
-                                             f'from_tags is True'
+            msg = ErrorNumbers.FB410.value + \
+                ' set_training_data: cannot specify a training_data argument if ' \
+                'from_tags is True'
             logger.critical(msg)
             raise FedbiomedValueError(msg)
+
         # case where no training data are passed
         if training_data is None:
             if from_tags is True:
                 if self._tags is None:
-                    msg = f"{ErrorNumbers.FB410.value}: attempting to set training data from undefined tags"
-                    logger.critical(msg)
-                    raise FedbiomedValueError(msg)
+                    raise FedbiomedValueError(
+                        f"{ErrorNumbers.FB410.value}: attempting to"
+                        "set training data from undefined tags. Please consider set tags before "
+                        "using set_tags method of the experiment.")
                 training_data = self._reqs.search(self._tags, self._nodes_filter)
             else:
                 self._fds = None
                 return None  # quick exit
+
         # from here, training_data is not None
         if isinstance(training_data, FederatedDataSet):
             self._fds = training_data
         elif isinstance(training_data, dict):
             self._fds = FederatedDataSet(training_data)
         else:
-            msg = ErrorNumbers.FB410.value + f' `training_data` has incorrect type: {type(training_data)}'
+            msg = ErrorNumbers.FB410.value + \
+                f' `training_data` has incorrect type: {type(training_data)}'
             logger.critical(msg)
             raise FedbiomedTypeError(msg)
         # check and ensure consistency
         if not self._fds_tags_consistent():
```

Comment:
@srcansiz @sharkovsky I don't feel very comfortable with `_fds_tags_consistent()`
I would propose to remove this function and have a simple rule for coherence (if we want to maintain some tags coherence):

```
if not from_tags:
  self._tags = None
```

Arguments for this proposal:
1. the actual behaviour is very difficult to undersand for the user

2. We might have to fix some cases

For example if:
- if `tags == [ 'one' ]` and  `training_data().data() == []`
- then `_fds_tags_consistent` returns False 
- so if `from_tags == False` then we set `self._tags = None` while we might consider the tags to be coherent with the datasets (any set of tags can give an empty set of datasets ...)

3. more generally, I suspect in some cases we cannot have a coherence check (or at least it questions what "coherent" means)

For example:
- if we have `self._tags [ 'un']`
- then we `set_training_data(FederatedDataset({ 'd1': { tags = [ 'un', 'deux'] } }), from_tags = False)
- so `_fds_tags_consistent` returns True as all `self._tags` are contained in `d1` tags
- BUT if we query `set_training_data(from_tags=True) we may get something like `{ 'd1': { tags = [ 'un', 'deux'] }, 'd2': { tags = [ 'un', 'trois'] } }` so it might not be coherent with the current status of our nodes and datasets. So I'm not sure the kind of coherence provided is very interesting ...

What do you think ?